core/src/main/java/org/elasticsearch/search/aggregations/AggregationBuilder.java

```diff
@@ -95,6 +96,23 @@ public String getName() {
     @Override
     public abstract AggregationBuilder subAggregations(AggregatorFactories.Builder subFactories);
 
+    public final AggregationBuilder rewrite(QueryRewriteContext context) throws IOException {
+        AggregationBuilder rewritten = doRewrite(context);
+        if (rewritten == this) {
```

Comment:
Agreed, though I put the loop in AggregatorFactories.Builder so that it loops over completely rewritten agg builders.